Sources/MobileSdk/BLEConnection.swift

```diff
@@ -0,0 +1,178 @@
+    private var totalBytesWritten = 0
+
+    /// Handle stream events.  Many of these we hand to local methods which the child classes are expected to
+    /// override.
+    func stream(_ aStream: Stream, handle eventCode: Stream.Event) {
+        switch eventCode {
+        case Stream.Event.openCompleted:
+            // TODO: This is a bit of a hack, but it'll do for now.  There are two streams, one input, one
+            // output, and we get notified about both.  We really only want to start doing things when
+            // both are available.
+            openCount += 1
+
+            if openCount == 2 {
+                streamIsOpen()
+            }
+
+        case Stream.Event.endEncountered:
+            openCount -= 1
+            streamEnded()
+
+        case Stream.Event.hasBytesAvailable:
+            streamBytesAvailable()
+            readBytes(from: aStream as! InputStream)
+
+        case Stream.Event.hasSpaceAvailable:
+            streamSpaceAvailable()
+            send()
+
+        case Stream.Event.errorOccurred:
+            streamError()
+
+        default:
+            streamUnknownEvent()
+        }
+    }
+
+    /// Public send() interface.
+    public func send(data: Data) {
+        outputData.append(data)
+        totalBytesWritten = 0
+        send()
+    }
+
+    /// Internal send() interface.
+    private func send() {
+        guard let ostream = channel?.outputStream, !outputData.isEmpty, ostream.hasSpaceAvailable else {
+            return
+        }
+        let bytesWritten = ostream.write(outputData)
+
+        totalBytesWritten += bytesWritten
+
+        // The isEmpty guard above should prevent div0 errors here.
+        let fracDone = Double(totalBytesWritten) / Double(outputData.count)
+
+        streamSentData(bytes: bytesWritten, total: totalBytesWritten, fraction: fracDone)
+
+        if bytesWritten < outputData.count {
+            outputData = outputData.advanced(by: bytesWritten)
+        } else {
+            outputData.removeAll()
+        }
+    }
+
+    /// Close the stream.
+    public func close() {
+        if let ch = channel {
+            ch.outputStream.close()
+            ch.inputStream.close()
+            ch.inputStream.remove(from: .main, forMode: .default)
+            ch.outputStream.remove(from: .main, forMode: .default)
+            ch.inputStream.delegate = nil
+            ch.outputStream.delegate = nil
+            openCount = 0
+        }
+
+        channel = nil
+    }
+
+    /// Read from the stream.
+    private func readBytes(from stream: InputStream) {
+        let bufLength = 1024
+        let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufLength)
+        defer {
+            buffer.deallocate()
+        }
+        let bytesRead = stream.read(buffer, maxLength: bufLength)
+        incomingData.append(buffer, count: bytesRead)
+
+        //    This is an awful hack to work around a hairy problem.  L2CAP is a stream protocol; there's
```

Comment:
In the L2CapDemo, it uses the `stateChangeCallback` event (https://github.com/paulw11/L2CapDemo/blob/master/L2CapDemo/CentralViewController.swift#L78), could it be applicable here? Maybe it's completely the wrong thing, but considering the reader does only one request, as does the holder, it could make sense to close the stream after each.

Reply:
The problem here is I don't know what we can count on 3rd party readers/holders playing nicely with closing the stream.  You'd be arguably within spec if you threw an exception on half the stream closing before the full exchange was complete.